### PR TITLE
PureScript Community Code of Conduct

### PR DESCRIPTION
**Pull request by thomashoneyman**
This is a draft of a new code of conduct for the PureScript community. Before commenting on this PR, please [read the RFC announcement on Discourse](https://discourse.purescript.org/t/rfc-purescript-community-code-of-conduct/2488).

### CODE REVIEW

File: CODE_OF_CONDUCT.md
Comment (thomashoneyman) [inline CODE_OF_CONDUCT.md]: We may also want to include that moderators can edit or delete a message, or ask that the person edits or deletes their message, if it's deemed something that shouldn't be preserved in the channel. This is a sensitive line to walk, but I suspect it will sometimes be necessary.

File: CODE_OF_CONDUCT.md
Comment (toastal) [inline CODE_OF_CONDUCT.md:35]: Is there a thematic breakup here to have 2 newlines? Typo or `---`?

Comment (thomashoneyman) [reply CODE_OF_CONDUCT.md:35]: This was unintentional -- it's just meant to be one newline.

File: CODE_OF_CONDUCT.md
Comment (hdgarrood) [inline CODE_OF_CONDUCT.md]: It appears that this only works for moderators who are currently available; I can't find a way of listing all the members of the moderators role including those who are not currently available.

File: CODE_OF_CONDUCT.md
Comment (rhendric) [inline CODE_OF_CONDUCT.md]: This group page doesn't exist or is private, says Discourse.

Comment (hdgarrood) [reply CODE_OF_CONDUCT.md]: It works for me, so it must only be visible to moderators. This is part of why I would prefer we list moderators explicitly with their names and email addresses here - I think making it easy to report incidents should be the priority, and listing email addresses seems like the easiest and most foolproof way to ensure that people can report incidents. Requiring people to navigate or sign up for some service they're not familiar with or for which the instructions might not work could be a very poor experience if you're already upset because of the incident you're trying to report, and not everyone is on Discord or Discourse.

**Comment by mikesol** [comment]
Looks great!

File: CODE_OF_CONDUCT.md
Comment (thomashoneyman) [reply CODE_OF_CONDUCT.md]: I did find that moderators can be found directly via the `about` page, which applies for everyone regardless of being logged in:
https://discourse.purescript.org/about

**Comment by JamieBallingall** [comment]
I have a bunch of minor notes that might help tighten up the language. But you may be working to a style guide that I don't know about.

EDIT: Specific comments moved to PR as per @JordanMartinez's suggestion below.

**Comment by JordanMartinez** [comment]
@JamieBallingall Could you add these comments to the part in the PR where it's stated? That will provide a bit more context.

File: CODE_OF_CONDUCT.md
Comment (JamieBallingall) [inline CODE_OF_CONDUCT.md]: Should the comma between "safe" and "inclusive" be replaced with an "and"? And perhaps replace "to" with "where we can". So would read "The PureScript community follows some simple ground rules to protect safe and inclusive spaces where we can discuss the PureScript programming language."

File: CODE_OF_CONDUCT.md
Comment (JamieBallingall) [inline CODE_OF_CONDUCT.md]: I think the comma before the "and" should be removed. Perhaps this is a writing style thing but all the other commas have some grammatical purpose while this one appears to indicate a pause.

File: CODE_OF_CONDUCT.md
Comment (JamieBallingall) [inline CODE_OF_CONDUCT.md]: Split "gender identity and expression" into "gender identity, gender expression". Similarly "religion, or sexual identity and orientation" becomes "religion, sexual identity or sexual orientation."

File: CODE_OF_CONDUCT.md
Comment (JamieBallingall) [inline CODE_OF_CONDUCT.md]: Replace "community leaders" with "moderators". Or define what we mean by "community leader".

File: CODE_OF_CONDUCT.md
Comment (JamieBallingall) [reply CODE_OF_CONDUCT.md]: I agree. Also, replace "community leaders" with "the moderators", remove the comma after leaders, replace "providing clarity around" with "clarifying", replace "an explanation of" with "explaining". 

So it now reads "A private, written warning from moderators clarifying the nature of the violation and explaining why the behavior was inappropriate."

File: CODE_OF_CONDUCT.md
Comment (JamieBallingall) [inline CODE_OF_CONDUCT.md]: I think the Causes should be in the same general format so people can see the escalation. So change to "A violation through a single serious incident or sustained inappropriate behavior".

File: CODE_OF_CONDUCT.md
Comment (JamieBallingall) [inline CODE_OF_CONDUCT.md]: I think it's important to make it clear that a permanent ban can be handed down for a single violation if that violation is serious enough. So change to "A violation through a single very serious incident or anything that demonstrates a pattern of violation of community standards".

File: CODE_OF_CONDUCT.md
Comment (JamieBallingall) [inline CODE_OF_CONDUCT.md]: Kinda trivial but all your other lists end with an "and". So perhaps this should be "We discuss what we can do to grow an inclusive, friendly, and respectful community."

File: CODE_OF_CONDUCT.md
Comment (JordanMartinez) [reply CODE_OF_CONDUCT.md]: I think either rendering (Thomas' current one and your suggested one) still say the same thing. So, I don't really have an opinion here.

File: CODE_OF_CONDUCT.md
Comment (JordanMartinez) [reply CODE_OF_CONDUCT.md]: The comma here could be removed or kept as-is. Again, I don't have an opinion.

File: CODE_OF_CONDUCT.md
Comment (JordanMartinez) [reply CODE_OF_CONDUCT.md]: I think those are intentionally grouped together.

File: CODE_OF_CONDUCT.md
Comment (JordanMartinez) [reply CODE_OF_CONDUCT.md]: I think "moderators" would be clearer. However, "community leaders" sounds more personal. Perhaps we should clarify how one can identify a "community leader" in the chatroom?

File: CODE_OF_CONDUCT.md
Comment (JordanMartinez) [reply CODE_OF_CONDUCT.md]: This is a good point. Ideally, most issues would go through levels of escalation and hopefully be addressed in lower-levels. However, it is possible that we could get a troll whose intent is to do harm. We wouldn't want to use this power frequently, but should still have it available if such a troll came.

File: CODE_OF_CONDUCT.md
Comment (JordanMartinez) [reply CODE_OF_CONDUCT.md]: I think this worth adding. It does make things a bit more readable.

File: CODE_OF_CONDUCT.md
Comment (JamieBallingall) [reply CODE_OF_CONDUCT.md]: I think that they probably were originally grouped together intentionally. But I believe splitting them is worthwhile for two reasons.
1. By grouping "gender identity and expression" and "sexual identity and orientation" we are saying that identity and expression in those instances are tied together in a way that, for example, race and color are not. That may be true -- it's a long way outside my field of expertise. But what we are doing with this sentence is highlighting examples and I think it would be better to be silent on the relationships between any of these attributes. We might also want to sort the list in alphabetical order to remove any hint of hierarchy between them.

2. No reasonable person would make the argument that "I was only harassing this person based on their sexual orientation, not their sexual identity. Therefore I was not harassing them based on their "sexual identity and sexual orientation" because 'false and true equals false'". But if there ever was a person who would make that argument it would be a troll with a background in computer science.

**Comment by thomashoneyman** [comment]
I’ve seen the great feedback here and on Discourse and I intend to write a new draft incorporating it this weekend. It’s been a busy week! I’ll ping everyone who has given advice so far when that commit is ready.

File: CODE_OF_CONDUCT.md
Comment (JamieBallingall) [reply CODE_OF_CONDUCT.md]: I agree. Is the escalation point worth making explicitly? I think it comes across fairly well just from the structure of the section.

File: CODE_OF_CONDUCT.md
Comment (ursi) [reply CODE_OF_CONDUCT.md]: I think this list should be removed altogether. You're never gonna get it right and it's logically the same statement if you remove it: "we pledge to make participation in our community a harassment-free experience for everyone."

Comment (hdgarrood) [reply CODE_OF_CONDUCT.md]: Strongly disagree. It may be the case that it’ll never be perfect but we can always amend it in the future if we need to. To me, if a CoC doesn’t actually list the protected characteristics like this, it’s a good signal that the community doesn’t recognise one or more of them as being worthy of protection, and this would make me _less_ likely to want to participate.

Comment (ursi) [reply CODE_OF_CONDUCT.md]: @hdgarrood but doesn't saying "harassment-free" mean no harassment of any kind? Sans the list, you think people will read this and think to themselves "hm, perhaps I am the unstated exception to this rule"?

Comment (rhendric) [reply CODE_OF_CONDUCT.md]: It's that different people have different definitions of what constitutes harassment. A list like this signals that the moderators are (or at least claim to be) aware of the specific concerns of the mentioned communities, and intend to take them into account when determining whether behavior merits moderator action. Sans the list, people will read this and wonder whether ‘harassment-free’ means that the specific, possibly controversial form of harassment that they've experienced or seen others experience is going to get a pass because it's not an overt threat.

Comment (ursi) [reply CODE_OF_CONDUCT.md]: Okay, this is a very good point. I still don't like the list as is, because to me it logically disguises it's true purpose: to more clearly define what the community considers harassment. Personally, I would prefer it be reworded/reorganized to be inline with this purpose.

**Comment by thomashoneyman** [comment]
I have made a significant update to the proposed code of conduct based on the comments made here and [on Discourse](https://discourse.purescript.org/t/rfc-purescript-community-code-of-conduct/2488/). I have largely attempted to protect the same things as before, but I have gone into more detail about what behavior is unwelcome and how moderators will respond in the case of an incident. I have also tried to make it friendlier and focused on shared respect, rather than focusing so much on disallowed behavior and the penalties you will face if you violate the code of conduct.

@toastal @rhendric @ursi @JamieBallingall I am not able to re-request review from you, but there's an update if you'd like to give this another read!

File: CODE_OF_CONDUCT.md
Comment (JordanMartinez) [inline CODE_OF_CONDUCT.md]: The line above says, "The following types of behavior are unacceptable..."

And now this line says "... it doesn't enumerate all possible forms of unwelcome behavior.."

The first line originally had a list that was deleted, so was the first line supposed to be deleted as well?

**Review by JordanMartinez — APPROVED** [review]
LGTM. Would there be another round of soliciting community feedback on this new draft before we finalize it?

**Comment by MonoidMusician** [comment]
Thanks Thomas! I was just thinking we should get back to this. Looking good to me.

File: CODE_OF_CONDUCT.md
Comment (thomashoneyman) [reply CODE_OF_CONDUCT.md]: I could do a better job merging them together. The point I'm trying to make is that this is a set of guidelines, but it is ultimately up to the discretion of moderators to decide what is acceptable even if it's not technically written out here. But perhaps this phrasing is verbose and unnecessary.

**Comment by thomashoneyman** [comment]
> LGTM. Would there be another round of soliciting community feedback on this new draft before we finalize it?

Yes, I think it should stay open for a while. @rhendric in particular had some great feedback and I'd like to wait for another review there. I'd also like to see if anyone on Discourse has more to say.

File: CODE_OF_CONDUCT.md
Comment (thomashoneyman) [reply CODE_OF_CONDUCT.md]: On a second look, I could just swap the two paragraphs

File: CODE_OF_CONDUCT.md
Comment (thomashoneyman) [reply CODE_OF_CONDUCT.md]: That's true, though you can still send a message to the group. We could have a brief meeting with current moderators and ask if everyone is OK with their name and email address being included in the document, or we could set up a `moderators@purescript.org` email address that forwards to moderator inboxes.

File: CODE_OF_CONDUCT.md
Comment (thomashoneyman) [reply CODE_OF_CONDUCT.md]: I've tried to reword things so that this sentiment (define some examples of what the community considers harassment) is more obvious.

**Review by MonoidMusician — APPROVED** [review]

**Review by garyb — APPROVED** [review]

File: CODE_OF_CONDUCT.md
Comment (rhendric) [inline CODE_OF_CONDUCT.md]:
```suggestion
* We discuss general functional programming concepts and other parts of programming, type theory, and mathematics that are relevant to PureScript.
```

File: CODE_OF_CONDUCT.md
Comment (rhendric) [inline CODE_OF_CONDUCT.md]: ‘If you are told’ by a moderator? By another member of the community? This is an important question; it's the difference between a community where unacceptable language is determined by the judgment of the moderators (and limited by their point-in-time awareness of the potential harm done by language), and a community where unacceptable language is defined as the union of everything found unacceptable by each member of the community.

If the latter is what is intended, you might want to rethink the language above about ‘arguing with someone who has pointed that out to you’. If the community norms get determined by the community, but the moderators don't permit the community to productively debate those norms, that would seem to create a sort of ratchet effect where it's trivial to establish new forbidden language and impossible to push back in the other direction, and I assume that's not how you actually intend to run things.

Just to emphasize, none of this is meant to be feedback about how you should and shouldn't moderate; I don't consider myself qualified to give that feedback. My primary concern here is that you clearly communicate your vision as moderators, with the intended audience being someone who would like to know in advance what they can expect before they invest in the community. My position is that a CoC ought to give as much detail as is practical on potentially thorny questions like this, not in order to bind the moderators into a legalistic framework, but rather in order to dispel ambiguity in the minds of the hesitant.

File: CODE_OF_CONDUCT.md
Comment (rhendric) [inline CODE_OF_CONDUCT.md]: ‘[C]ontact the perpetrator to verify what has transpired’ sounds like maybe the ‘perpetrator’ (alleged perpetrator? accused? complainee?) has a chance to give their side of the story and talk their way out of the following consequences, but the rest of the language of this paragraph doesn't imply that it's conditional on the report being verified. Here are a few example edits that I think would clarify this, depending on where you intend to fall on the believe-all-reports/benefit-of-the-doubt spectrum:

```suggestion
If we witness or receive a report about abusive behavior, we will contact the perpetrator to inform them of the observation or report. They will be removed from the PureScript community. Their Discord and Discourse accounts will be deactivated and they will be removed from PureScript GitHub organizations.
```

```suggestion
If we witness or receive a report about abusive behavior, we will contact the complainee to verify what has transpired. At the time of the report, the complainee will be removed from the PureScript community. Their Discord and Discourse accounts will be deactivated and they will be removed from PureScript GitHub organizations. If the complainee contests the facts of the report, we will investigate further and, if in the judgment of the moderation team no abusive behavior took place, the complainee's accounts will be reinstated with no further consequences.
```

```suggestion
If we witness or receive a report about abusive behavior, we will contact the complainee and investigate the details of the incident. In the event that the incident constitutes abusive behavior in the opinion of the moderation team, the perpetrator will be removed from the PureScript community. Their Discord and Discourse accounts will be deactivated and they will be removed from PureScript GitHub organizations.
```

**Review by rhendric — APPROVED** [review]
Personally I like the changes; there are a couple of things I think could be clearer but well done in general.

File: CODE_OF_CONDUCT.md
Comment (thomashoneyman) [reply CODE_OF_CONDUCT.md]: > ‘If you are told’ by a moderator

This is the correct reading. I'd assume that in practice, for particularly obvious cases, members of the community might note some language or action isn't generally considered acceptable here. But it's only when you are told _by a moderator_ that the rule about not arguing comes into play.

There is no intent to empower individuals to arbitrarily decide on new forbidden language or behavior, if only because it then becomes extremely difficult to moderate. But individuals can certainly suggest or ask for certain things to be moderated, in which case it is still up to the moderators to make any 'official' decision on it.

Comment (thomashoneyman) [reply CODE_OF_CONDUCT.md]: I've updated this! Thanks for pointing it out.

**Comment by thomashoneyman** [comment]
This now has sufficient approval by the core team to be merged, but I'll still keep it open a little longer in case anyone has remaining thoughts.

**Pull request merged**
